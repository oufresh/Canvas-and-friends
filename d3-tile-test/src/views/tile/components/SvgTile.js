import React from 'react';
import { TILE_SIZE } from '../../../modules/maps/utils/constants';

type SvgTilePropsType = {
    key: string,
    x: number,
    y: number,
    z: number,
    svgContent?: string,
    urlSvgProvider?: Function,
    getTilePromise?: Promise,
    opts?: any,
    baseUrl?: string,
    width?: number,
    height?: number
};

export const defaultHeight: number = TILE_SIZE;
export const defaultWidth: number = TILE_SIZE;

const createSvgMarkup = (svg: string) => {
    return { __html: svg };
}

class SvgTile extends React.PureComponent<SvgTilePropsType>
{
    constructor(props: SvgTilePropsType) 
    {
        super(props);
        this.state = {
            svgString: this.props.svgContent ? this.props.svgContent : ''
        }
    }

    componentDidMount()
    {
        //console.log("SvgTile componentDidMount");
        if (this.props.urlSvgProvider) {
            const url = this.props.urlSvgProvider(this.props.z, this.props.x, this.props.y);
            fetch(url).then(r => {
                if (r.ok === true)
                    return r.text();
                else
                    throw new Error(r.statusText)
            }).then(svgString => {
                //console.log('Received tile: ' + svgString);
                this.setState({
                    svgString
                });
            }).catch(e => {
                console.error(e);
            });
        }
        else if (this.props.getTilePromise) {
            this.props.getTilePromise(this.props.baseUrl, this.props.z, this.props.x, this.props.y, this.props.renderInfo).then(svgString => {
                //console.log('Received tile: ' + svgString);
                this.setState({
                    svgString
                });
            }).catch(e => {
                console.error(e);
            });
        }
    }

    render()
    {
        const h = this.props.height ? this.props.height : defaultHeight;
        const w = this.props.width ? this.props.width : defaultWidth;
        const content = createSvgMarkup(this.state.svgString);
        const transform = 'translate('+ this.props.x*w+' '+ this.props.y*h+')';
    
        return <g transform={transform} width={w} height={h} dangerouslySetInnerHTML={content} />;
    }
}

export default SvgTile;