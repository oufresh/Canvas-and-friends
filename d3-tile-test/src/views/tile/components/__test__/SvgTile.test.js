//@flow
import React from 'react';
import { shallow } from 'enzyme';
import SvgTile from '../SvgTile';

describe('SvgTile Component', () => {

    it('renders without a crash with preloaded svg content', () => {
        const component = shallow(<SvgTile key="test" x={0} y={0} z={0} svgContent={'<rect stroke="#000" id="svgtest" height="254" width="254" y="1" x="1" stroke-width="1.5" fill="rgba(0,0,0,0)"></rect>'}/>);
        expect(component).toMatchSnapshot();
    });

});