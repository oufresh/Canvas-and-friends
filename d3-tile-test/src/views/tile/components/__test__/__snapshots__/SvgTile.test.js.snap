// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`SvgTile Component renders without a crash with preloaded svg content 1`] = `
ShallowWrapper {
  Symbol(enzyme.__root__): [Circular],
  Symbol(enzyme.__unrendered__): <SvgTile
    svgContent="<rect stroke=\\"#000\\" id=\\"svgtest\\" height=\\"254\\" width=\\"254\\" y=\\"1\\" x=\\"1\\" stroke-width=\\"1.5\\" fill=\\"rgba(0,0,0,0)\\"></rect>"
    x={0}
    y={0}
    z={0}
  />,
  Symbol(enzyme.__renderer__): Object {
    "batchedUpdates": [Function],
    "getNode": [Function],
    "render": [Function],
    "simulateEvent": [Function],
    "unmount": [Function],
  },
  Symbol(enzyme.__node__): Object {
    "instance": null,
    "key": undefined,
    "nodeType": "host",
    "props": Object {
      "dangerouslySetInnerHTML": Object {
        "__html": "<rect stroke=\\"#000\\" id=\\"svgtest\\" height=\\"254\\" width=\\"254\\" y=\\"1\\" x=\\"1\\" stroke-width=\\"1.5\\" fill=\\"rgba(0,0,0,0)\\"></rect>",
      },
      "height": 256,
      "transform": "translate(0 0)",
      "width": 256,
    },
    "ref": null,
    "rendered": null,
    "type": "g",
  },
  Symbol(enzyme.__nodes__): Array [
    Object {
      "instance": null,
      "key": undefined,
      "nodeType": "host",
      "props": Object {
        "dangerouslySetInnerHTML": Object {
          "__html": "<rect stroke=\\"#000\\" id=\\"svgtest\\" height=\\"254\\" width=\\"254\\" y=\\"1\\" x=\\"1\\" stroke-width=\\"1.5\\" fill=\\"rgba(0,0,0,0)\\"></rect>",
        },
        "height": 256,
        "transform": "translate(0 0)",
        "width": 256,
      },
      "ref": null,
      "rendered": null,
      "type": "g",
    },
  ],
  Symbol(enzyme.__options__): Object {
    "adapter": ReactSixteenAdapter {
      "options": Object {
        "enableComponentDidUpdateOnSetState": true,
        "lifecycles": Object {
          "componentDidUpdate": Object {
            "onSetState": true,
          },
          "getSnapshotBeforeUpdate": true,
        },
      },
    },
    "attachTo": undefined,
    "hydrateIn": undefined,
  },
}
`;
